client/lib/wpcom-undocumented/lib/undocumented.js

```diff
@@ -844,17 +844,21 @@ Undocumented.prototype.getTitanMailboxAvailability = function ( domain, mailbox,
 /**
  * Get a list of WordPress.com products
  *
- * @param {Function} fn The callback function
+ * @param {object|Function}	queryOrCallback
+ *   Either an object with query string parameters, or a callback function
  */
-Undocumented.prototype.getProducts = function ( fn ) {
+Undocumented.prototype.getProducts = function ( queryOrCallback ) {
 	debug( '/products query' );
-	return this._sendRequest(
-		{
-			path: '/products',
-			method: 'get',
-		},
-		fn
-	);
+
+	if ( 'function' === typeof queryOrCallback ) {
+		// The passed-in argument is a callback;
+		// make sure to pass in a safe query param object
+		return this.wpcom.req.get( '/products', {}, queryOrCallback );
+	}
+
+	// At this point, `queryOrCallback` should either be falsy,
+	// or it should be a query param object
+	return this.wpcom.req.get( '/products', queryOrCallback || {} );
```

Comment:
This makes the `getProducts` function behave a bit oddly:
- `getProducts( { type } ).then( ... )` works OK
- `getProducts( ( error, data ) => { ... } )` also works OK
- but `getProducts( { type }, ( error, data ) => { ... } )` never calls the callback.

In other words, the callback and promise forms are not equal interchangeable.

The `sharingButtons` function in this file is a good blueprint how to correctly do an optional `query` argument followed by a callback.

Reply:
I wouldn't mind if we removed the `wpcom.undocumented().getProducts()` completely and just called
```js
wpcom.req.get( '/products', { type } )
```
directly at the two places that use it. It supports all query/callback combinations out of the box:
```js
wpcom.req.get( '/products' ).then();
wpcom.req.get( '/products', { type } ).then();
wpcom.req.get( '/products', cb );
wpcom.req.get( '/products', { type }, cb );
```